### ISSUE TITLE
Connection spins when attempting to connect through gateway server

### ISSUE THREAD

**Issue by jackdozier** [FreeRDP/FreeRDP#3185]
Using FreeRDP version 2.0.0-dev (git 9afa928)
CentOS Linux release 7.2.1511 (Core) x86_64

Trying to connect to Windows 10 behind a gateway. I'm not sure of the version of the gateway server. Software prompts for gateway password and performs handshake. Connection hangs with high cpu usage after debug output says "Tunnel response received". (Full command line and output below.)

Narrowed issue to libfreerdp/core/gateway/rdg.c. These two commits looked related to issue, so I reverted to commit 48c7a8dc1d334bc316f8de466ce94ff91e368e92 and the issue went away.

> commit 106479d6f27cdd34a8d7dd8e654d890542ab1fcb
> Author: Armin Novak <armin.novak@thincast.com>
> Date:   Tue Jan 26 09:34:22 2016 +0100
> 
>     Fixed clientNameLen initialization.
> 
> commit 73ec3d6acaea4ca69dfd43df3cb847cc83310324
> Author: Armin Novak <armin.novak@thincast.com>
> Date:   Thu Jan 21 14:47:10 2016 +0100
> 
>     Removed fixed size strings.

Not wanting to throw away Armin's work, I reverted back to the latest code and applied the [attached](https://github.com/FreeRDP/FreeRDP/files/156145/0001-Fixed-packet-size-issue.txt) changes and it continued to work for me. Basically, removed the unnecessary(?) zero being written to the stream and updated the packet size. I didn't look into the protocol specs, so I have no idea if that zero is necessary. I did test it on two other environments behind gateway servers, though, and they both worked with my changes.

xfreerdp /g:remote.example.com /v:client.example.com /gu:user /gd:example.com /u:user /d:example.com /log-level:DEBUG
[21:19:31:461] [17612:17613] [DEBUG][com.freerdp.client.x11] - Searching for XInput pointer device
[21:19:31:461] [17612:17613] [DEBUG][com.freerdp.client.x11] - Pointer device: 11
[21:19:31:462] [17612:17613] [DEBUG][com.freerdp.core.nego] - Enabling security layer negotiation: TRUE
[21:19:31:462] [17612:17613] [DEBUG][com.freerdp.core.nego] - Enabling restricted admin mode: FALSE
[21:19:31:462] [17612:17613] [DEBUG][com.freerdp.core.nego] - Enabling RDP security: TRUE
[21:19:31:462] [17612:17613] [DEBUG][com.freerdp.core.nego] - Enabling TLS security: TRUE
[21:19:31:462] [17612:17613] [DEBUG][com.freerdp.core.nego] - Enabling NLA security: TRUE
[21:19:31:462] [17612:17613] [DEBUG][com.freerdp.core.nego] - Enabling NLA extended security: FALSE
[21:19:31:462] [17612:17613] [DEBUG][com.freerdp.core.nego] - state: NEGO_STATE_NLA
[21:19:31:462] [17612:17613] [DEBUG][com.freerdp.core.nego] - Attempting NLA security
GatewayPassword: 
[21:19:35:566] [17612:17613] [DEBUG][com.winpr.sspi] - InitSecurityInterfaceExA
[21:19:35:743] [17612:17613] [DEBUG][com.freerdp.core.gateway.rdg] - Out Channel authorization required
[21:19:35:944] [17612:17613] [DEBUG][com.freerdp.core.gateway.rdg] - Out Channel authorization complete
[21:19:36:417] [17612:17613] [DEBUG][com.winpr.sspi] - InitSecurityInterfaceExA
[21:19:36:557] [17612:17613] [DEBUG][com.freerdp.core.gateway.rdg] - In Channel authorization required
[21:19:36:645] [17612:17613] [DEBUG][com.freerdp.core.gateway.rdg] - In Channel authorization complete
[21:19:36:789] [17612:17613] [DEBUG][com.freerdp.core.gateway.rdg] - Handshake response received
[21:19:36:927] [17612:17613] [DEBUG][com.freerdp.core.gateway.rdg] - Tunnel response received

**Comment by akallabeth** [comment]
@jackdozier nice find, looks like I messed up the size of the packet. will have a look into that!

**Comment by hardening** [comment]
I'm quite bothered that a bug in packet forgery causes 100% CPU, this smells like some piece of code is doing busy waiting somewhere, isn't it ?

**Comment by akallabeth** [comment]
@hardening True, this smells like the error handling needs some tweaks

**Comment by jackdozier** [comment]
@akallabeth Newest git pull works for me. (git 167fd47) Thanks!

@hardening @akallabeth I was concerned by the 100% CPU too and narrowed it down to the rdg_receive_packet function. It seems to be continuously looping in the while loop polling BIO_read, getting a status of zero and then repeating. At this point, I assume the server has dropped or reset the connection, but the client is never going to know about it. I'm not sure if there's a blocking variant of BIO_read or if some wait/timeout algorithm needs to be devised (or better yet, why doesn't openssl detect the reset and pass a negative status?). Just thought I'd pass along what I saw.

**Comment by giox069** [comment]
Just to add my comments on "busy loops": some months ago I discovered one. Take a look at `master/libfreerdp/core/connection.c`, the following lines:
```
while (rdp->state != CONNECTION_STATE_ACTIVE)
	{
		if (rdp_check_fds(rdp) < 0)
		{
			if (!freerdp_get_last_error(rdp->context))
				freerdp_set_last_error(rdp->context, FREERDP_ERROR_CONNECT_TRANSPORT_FAILED);
			return FALSE;
		}
	}
```
I did not tested it recently... but I can think rdp_check_fds(rdp) is a nonblocking function, and thus this is a busy loop. I remember I have placed a printf inside and seen some hundred of output lines...

**Comment by jackdozier** [comment]
I found other examples of BIO_read in libfreerdp/core/gateway/rdg.c and they call a BIO_should_retry function to decide whether it's a temporary or permanent error. I took a stab at adding the functionality to the two BIO_reads in the rdg_receive_packet function. It required adding an else statement to rdg_out_channel_recv. It stops the busy loop on my test case with the malformed packet.

[0001-Added-error-checking-after-BIO_read-to-decide-whethe.txt](https://github.com/FreeRDP/FreeRDP/files/157436/0001-Added-error-checking-after-BIO_read-to-decide-whethe.txt)


Not sure if you have a better way of handling the error, so I won't be offended if you don't use my contribution. I have not looked at the issue @giox069 mentioned.
